core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -824,9 +827,15 @@ private[deploy] class Master(
         // If the cores left is less than the coresPerExecutor,the cores left will not be allocated
         if (app.coresLeft >= coresPerExecutor) {
           // Filter out workers that don't have enough resources to launch an executor
-          val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
+          val aliveWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
             .filter(canLaunchExecutor(_, resourceDesc))
-            .sortBy(_.coresFree).reverse
+          val usableWorkers = workerSelectionPolicy match {
+            case CORES_FREE_ASC => aliveWorkers.sortBy(w => (w.coresFree, w.id))
+            case CORES_FREE_DESC => aliveWorkers.sortBy(w => (w.coresFree, w.id)).reverse
```

Comment:
Hmm, this seems to sort workers by `(free cores, id)`? Shouldn't we sort them by free cores only?

Reply:
Ya, the ID is added to be deterministic. Previously, it's based on HashSet and indeterministic among `coresFree`.